In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')
from sklearn.preprocessing import OneHotEncoder,MultiLabelBinarizer,LabelEncoder,MinMaxScaler,StandardScaler, RobustScaler
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import ADASYN
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense,Dropout
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.models import load_model
from sklearn.utils import class_weight
from sklearn.metrics import classification_report
import numpy as np

In [5]:
data=pd.read_csv('Lending_Loan_Company_Data_Analysis.csv')
data_num=data.select_dtypes(exclude='object')

# Feature Engineering

## Check for Collinearity B/W the Columns

In [6]:
for i in data_num.columns:
    corr=data_num.corr()
    corr[i].sort_values(ascending=False)
    print(corr)

                           Credit_Policy  Interest_Rate  Monthly_Installment  \
Credit_Policy                   1.000000      -0.294089             0.058770   
Interest_Rate                  -0.294089       1.000000             0.276140   
Monthly_Installment             0.058770       0.276140             1.000000   
Natural_log_Annual_Income       0.034906       0.056383             0.448102   
Debt_Income_Ratio              -0.090901       0.220006             0.050202   
FICO_Score                      0.348319      -0.714821             0.086039   
Days_with_Credit_Line           0.099026      -0.124022             0.183297   
Revolving_Balance              -0.187518       0.092527             0.233625   
Revolving_Utilization          -0.104095       0.464837             0.081356   
Enquiries_Last_6_Months        -0.535511       0.202780            -0.010419   
Deliquency_Last_2_Yrs          -0.076318       0.156079            -0.004368   
Public_Derogatory_Records      -0.054243

### Since, Most of the Columns have the Scores less Than 0.5 between them, There is no need to drop off the columns

# Model Building

## 1. Encode the Labels

In [7]:
data_model=data.copy()

In [8]:
le=LabelEncoder()

In [9]:
data_model

,Credit_Policy,Loan_Purpose,Interest_Rate,Monthly_Installment,Natural_log_Annual_Income,Debt_Income_Ratio,FICO_Score,Days_with_Credit_Line,Revolving_Balance,Revolving_Utilization,Enquiries_Last_6_Months,Deliquency_Last_2_Yrs,Public_Derogatory_Records,Paid_Status
0,1,debt_consolidation,0.1189,829.10,11.350407,19.48,737,5639.958333,28854,52.1,0,0,0,Paid
1,1,credit_card,0.1071,228.22,11.082143,14.29,707,2760.000000,33623,76.7,0,0,0,Paid
2,1,debt_consolidation,0.1357,366.86,10.373491,11.63,682,4710.000000,3511,25.6,1,0,0,Paid
3,1,debt_consolidation,0.1008,162.34,11.350407,8.10,712,2699.958333,33667,73.2,1,0,0,Paid
4,1,credit_card,0.1426,102.92,11.299732,14.97,667,4066.000000,4740,39.5,0,1,0,Paid
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9573,0,all_other,0.1461,344.76,12.180755,10.39,672,10474.000000,215372,82.1,2,0,0,Unpaid
9574,0,all_other,0.1253,257.70,11.141862,0.21,722,4380.000000,184,1.1,5,0,0,Unpaid
9575,0,debt_consolidation,0.1071,97.81,10.596635,13.09,687,3450.041667,10036,82.9,8,0,0,Unpaid
9576,0,home_improvement,0.1600,351.58,10.819778,19.18,692,1800.000000,0,3.2,5,0,0,Unpaid


In [10]:
data['Loan_Purpose'].unique()

array(['debt_consolidation', 'credit_card', 'all_other',
       'home_improvement', 'small_business', 'major_purchase',
       'educational'], dtype=object)

In [11]:
data_model['Loan_Purpose'].unique()

array(['debt_consolidation', 'credit_card', 'all_other',
       'home_improvement', 'small_business', 'major_purchase',
       'educational'], dtype=object)

In [12]:
data

,Credit_Policy,Loan_Purpose,Interest_Rate,Monthly_Installment,Natural_log_Annual_Income,Debt_Income_Ratio,FICO_Score,Days_with_Credit_Line,Revolving_Balance,Revolving_Utilization,Enquiries_Last_6_Months,Deliquency_Last_2_Yrs,Public_Derogatory_Records,Paid_Status
0,1,debt_consolidation,0.1189,829.10,11.350407,19.48,737,5639.958333,28854,52.1,0,0,0,Paid
1,1,credit_card,0.1071,228.22,11.082143,14.29,707,2760.000000,33623,76.7,0,0,0,Paid
2,1,debt_consolidation,0.1357,366.86,10.373491,11.63,682,4710.000000,3511,25.6,1,0,0,Paid
3,1,debt_consolidation,0.1008,162.34,11.350407,8.10,712,2699.958333,33667,73.2,1,0,0,Paid
4,1,credit_card,0.1426,102.92,11.299732,14.97,667,4066.000000,4740,39.5,0,1,0,Paid
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9573,0,all_other,0.1461,344.76,12.180755,10.39,672,10474.000000,215372,82.1,2,0,0,Unpaid
9574,0,all_other,0.1253,257.70,11.141862,0.21,722,4380.000000,184,1.1,5,0,0,Unpaid
9575,0,debt_consolidation,0.1071,97.81,10.596635,13.09,687,3450.041667,10036,82.9,8,0,0,Unpaid
9576,0,home_improvement,0.1600,351.58,10.819778,19.18,692,1800.000000,0,3.2,5,0,0,Unpaid


## Description of Labels

### 1. Loan Purpose
#### * debt_consolidation - 2
#### * credit_card - 1
#### * all_other - 0
#### * home_improvement -4
#### * small_business - 6
#### * major_purchase - 5
#### * educational - 3

### 2. Paid Status
#### * Paid - 0
#### * Unpaid - 1

In [13]:
data_model['Installment_to_Income'] = data_model['Monthly_Installment'] / (data['Natural_log_Annual_Income'] / 12)
data_model['Interest_Installment'] = data_model['Interest_Rate'] * data_model['Monthly_Installment']
data_model['Revolving_to_Income'] = data_model['Revolving_Balance'] / data_model['Natural_log_Annual_Income']
data_model['Credit_Length_Years'] = data_model['Days_with_Credit_Line'] / 365
data_model['Risk_Score'] = data_model['Interest_Rate'] / data_model['FICO_Score']

In [14]:
x=data_model.drop(['Paid_Status','Natural_log_Annual_Income'],axis=1)
y=le.fit_transform(data_model['Paid_Status'])

In [15]:
xtrain,xtemp,ytrain,ytemp=train_test_split(x,y,test_size=0.3,random_state=42)
xeval,xtest,yeval,ytest=train_test_split(xtemp,ytemp,test_size=0.5,random_state=42)

In [16]:

cat_cols = ['Loan_Purpose','Credit_Policy', 'Loan_Purpose', 'Enquiries_Last_6_Months' ,'Deliquency_Last_2_Yrs','Public_Derogatory_Records']

# Outlier numeric
robust_cols = [
    'Interest_Rate',
    'Monthly_Installment',
    'FICO_Score',
    'Days_with_Credit_Line',
    'Revolving_Balance',
    'Installment_to_Income',
    'Interest_Installment'
    
]

# Remaining numeric
standard_cols = [col for col in xtrain.columns 
                 if col not in robust_cols + cat_cols]



In [17]:

preprocessor = ColumnTransformer(
    transformers=[
        ('robust', RobustScaler(), robust_cols),
        ('standard', StandardScaler(), standard_cols),
        ('cat',OneHotEncoder(handle_unknown='ignore'), cat_cols)   
    ]
)

# Step 1: Fit scaler on original train
X_train_scaled = preprocessor.fit_transform(xtrain)

# Step 2: Apply same transform
X_eval = preprocessor.transform(xeval)
X_test = preprocessor.transform(xtest)


In [29]:

class_weight = {0:1, 1:5}


In [30]:
model = Sequential()

# Input Layer + Hidden Layer 1
model.add(Dense(64, activation='relu', input_dim=X_train_scaled.shape[1]))
model.add(Dropout(0.3))

# Hidden Layer 2
model.add(Dense(32, activation='relu'))
model.add(Dropout(0.2))

# Hidden Layer 3
model.add(Dense(16, activation='relu'))

# Output Layer
model.add(Dense(1, activation='sigmoid'))  # Binary classification

# Compile
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['Precision', 'Recall']
)

# Early Stopping (VERY IMPORTANT)
early_stop = EarlyStopping(
    monitor='val_loss',
    patience=20,
    restore_best_weights=True
)

# Train
history = model.fit(
    X_train_scaled, ytrain,
    validation_data=(X_eval, yeval),
    class_weight=class_weight,
    epochs=200,
    batch_size=32,
    callbacks=[early_stop]
    
)

Epoch 1/200
210/210 ━━━━━━━━━━━━━━━━━━━━ 12s 21ms/step - Precision: 0.2219 - Recall: 0.4817 - loss: 1.1079 - val_Precision: 0.2246 - val_Recall: 0.7064 - val_loss: 0.6864
Epoch 2/200
210/210 ━━━━━━━━━━━━━━━━━━━━ 3s 14ms/step - Precision: 0.2320 - Recall: 0.5858 - loss: 1.0749 - val_Precision: 0.2303 - val_Recall: 0.5234 - val_loss: 0.6135
Epoch 3/200
210/210 ━━━━━━━━━━━━━━━━━━━━ 4s 16ms/step - Precision: 0.2313 - Recall: 0.5698 - loss: 1.0760 - val_Precision: 0.2482 - val_Recall: 0.5915 - val_loss: 0.6383
Epoch 4/200
210/210 ━━━━━━━━━━━━━━━━━━━━ 3s 13ms/step - Precision: 0.2394 - Recall: 0.5951 - loss: 1.0612 - val_Precision: 0.2490 - val_Recall: 0.5106 - val_loss: 0.6061
Epoch 5/200
210/210 ━━━━━━━━━━━━━━━━━━━━ 3s 12ms/step - Precision: 0.2412 - Recall: 0.6111 - loss: 1.0570 - val_Precision: 0.2418 - val_Recall: 0.4085 - val_loss: 0.5902
Epoch 6/200
210/210 ━━━━━━━━━━━━━━━━━━━━ 3s 12ms/step - Precision: 0.2454 - Recall: 0.5933 - loss: 1.0545 - val_Precision: 0.2505 - val_Recall: 0.587

In [31]:
loss, precision, recall = model.evaluate(X_eval, yeval)
print("Loss:", loss)
print("Precision:", precision)
print("Recall:", recall)

45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 12ms/step - Precision: 0.2500 - Recall: 0.4043 - loss: 0.5764
Loss: 0.5763811469078064
Precision: 0.25
Recall: 0.40425533056259155


In [34]:
ypred=model.predict(X_eval)

ypred_t = (model.predict(X_eval) > 0.45).astype(int)

print(classification_report(yeval, ypred_t))


45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 12ms/step
45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step
              precision    recall  f1-score   support

           0       0.89      0.62      0.73      1202
           1       0.24      0.61      0.34       235

    accuracy                           0.62      1437
   macro avg       0.57      0.62      0.54      1437
weighted avg       0.78      0.62      0.67      1437



In [35]:
ypred_Test=model.predict(X_test)

ypred_test = (model.predict(X_test) > 0.45).astype(int)

print(classification_report(ytest, ypred_test))

45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step
45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step
              precision    recall  f1-score   support

           0       0.90      0.63      0.74      1206
           1       0.25      0.63      0.36       231

    accuracy                           0.63      1437
   macro avg       0.57      0.63      0.55      1437
weighted avg       0.79      0.63      0.68      1437



#                                      Insights

1. Handling Class Imbalance

The dataset exhibited class imbalance, with significantly fewer 'Unpaid' (defaulter) cases compared to 'Paid' loans. Initially, ADASYN was applied to generate synthetic minority-class samples. Although this improved recall, it also increased the number of false positives, resulting in low precision.

To address this issue, class weights were introduced during model training. This approach improved the balance between precision and recall by assigning higher penalties to misclassification of defaulters, without generating synthetic data that could introduce noise.

2. Comparison of ADASYN and Class Weights

A comparison was performed between ADASYN, class weights, and their combination:

ADASYN alone significantly improved recall by increasing detection of defaulters, but it reduced precision due to excessive prediction of the minority class.
Class weights alone provided more stable and balanced performance by improving recall while maintaining better precision.
Combining ADASYN with class weights caused overcompensation toward the minority class, resulting in excessive false positives and degraded overall performance.

Therefore, class weights alone were selected as the optimal imbalance-handling strategy for the Lending Club loan default prediction problem.

3. Threshold Tuning

The default classification threshold of 0.5 resulted in low recall for the 'Unpaid' class, meaning that many risky borrowers were not identified by the model.

To improve detection of defaulters, threshold tuning was performed:

Lower thresholds (such as 0.3) increased recall but caused a significant drop in precision due to excessive false positives.
Higher thresholds (0.6–0.7) improved precision but failed to detect many defaulters, resulting in very low recall.
A threshold of 0.45 provided the best trade-off between precision and recall.

This demonstrates the importance of threshold tuning in imbalanced financial risk prediction problems.

4. Final Model Performance

The final Artificial Neural Network (ANN) model was trained using class weights and evaluated using a classification threshold of 0.45.

The model achieved approximately:

Recall ≈ 63% for the 'Unpaid' class, indicating that the model successfully identified a majority of risky borrowers.
Precision ≈ 25%, suggesting the presence of moderate false positives, which is acceptable in risk-sensitive lending applications where detecting defaulters is more important.
F1-score ≈ 0.36, reflecting a reasonable balance between precision and recall for the minority class.

Additionally, similar performance on both evaluation and test datasets indicates that the model generalized well and did not suffer from significant overfitting.

5. Business Interpretation

The primary objective of the project was to identify potentially risky loan applicants and reduce financial loss caused by loan defaults.

Since missing a defaulter can be more costly than incorrectly flagging a safe borrower, the model was optimized toward improving recall for the minority class. The final model therefore acts as an effective risk-screening system for identifying high-risk customers in lending applications.

# Summary

From a business perspective, the model prioritizes the identification of defaulters (Unpaid customers), since failing to detect such borrowers can result in significant financial losses for the lending institution.

To improve detection of risky applicants, the decision threshold was lowered to 0.45, allowing the model to classify moderately risky customers as potential defaulters. Although this approach increases false positives, it substantially reduces the likelihood of approving high-risk loan applicants.

This trade-off is consistent with real-world financial risk management practices, where reducing default risk is generally more important than maximizing loan approval rates.